File: editor/en/[340]Longest Substring with At Most K Distinct Characters.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        count_count = {}
        left = 0
        result = 0

        for right in range(len(s)):
            char = s[right]
            count_count[char] = count_count.get(char, 0) + 1
            while len(count_count) > k:
                ltft_char = s[left]
                count_count[ltft_char] -= 1
                if count_count[ltft_char] == 0:
                    del count_count[ltft_char]
                left += 1
            result = max(result, right - left + 1)

        return result
# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[340]Longest Substring with At Most K Distinct Characters.java
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int[] arr = new int[128];
        int left = 0;
        int res = 0;
        int count = 0;

        for (int right = 0; right < s.length(); right++) {
            arr[s.charAt(right)]++;

            if (arr[s.charAt(right)] == 1) {
                count++;
            }

            while (count > k) {
                if (arr[s.charAt(left)] == 1) {
                    count--;
                }

                arr[s.charAt(left)]--;
                left++;
            }

            res = Math.max(res, right - left + 1);
        }

        return res;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
